File: 383-ransom-note/solution.cpp
#include <string>
#include <unordered_map>

class Solution
{
private:
  static std::unordered_map<char, size_t> create_counter(const std::string &s)
  {
    std::unordered_map<char, size_t> counter;
    for (const auto c : s)
    {
      ++counter[c];
    }

    return counter;
  }

public:
  bool canConstruct(std::string ransomNote, std::string magazine)
  {
    const auto ransom_counter = Solution::create_counter(ransomNote);
    const auto mag_counter = Solution::create_counter(magazine);

    for (const auto &letter_count : ransom_counter)
    {
      const auto it = mag_counter.find(letter_count.first);
      if (it == mag_counter.end())
      {
        return false;
      }

      if (letter_count.second > it->second)
      {
        return false;
      }
    }

    return true;
  }
};

File: 383-ransom-note/solution.py
from collections import Counter


class Solution:
  def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    ransom_counter = Counter(ransomNote)
    magazine_counter = Counter(magazine)

    for letter, count in ransom_counter.items():
      if letter not in magazine_counter:
        return False

      mag_count = magazine_counter[letter]
      if count > mag_count:
        return False

    return True
